src/agent/proxy.c

```diff
@@ -0,0 +1,135 @@
+#include "libhirte/log/log.h"
+
+#include "agent.h"
+#include "proxy.h"
+
+
+static int proxy_job_method_set_ready(sd_bus_message *m, void *userdata, sd_bus_error *ret_error);
+
+static const sd_bus_vtable proxy_job_vtable[] = {
+        SD_BUS_VTABLE_START(0), SD_BUS_METHOD("Ready", "s", "", proxy_job_method_set_ready, 0), SD_BUS_VTABLE_END
+};
+
+static int proxy_job_method_set_ready(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        const char *status = NULL;
+        int r = sd_bus_message_read(m, "s", &status);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+        }
+
+        r = 0;
+        UnitActiveState active_state = active_state_from_string(status);
+        if (active_state == UNIT_ACTIVE) {
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_return(proxy->request_message, "");
+        } else if (active_state == UNIT_FAILED) {
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_errorf(
+                                proxy->request_message, SD_BUS_ERROR_FAILED, "Proxy service failed to start");
+        }
+
+        if (r < 0) {
+                hirte_log_errorf("Failed to sent ready status message to proxy: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to set ready status");
+        }
+
+        return sd_bus_reply_method_return(m, "");
+}
+
+
+ProxyJob *
+                proxy_job_new(Agent *agent,
+                              const char *node,
+                              const char *unit,
+                              const char *mode,
+                              sd_bus_message *request_message) {
+        static uint32_t next_id = 0;
+
+        _cleanup_proxy_job_ ProxyJob *proxy = malloc0(sizeof(ProxyJob));
+        if (proxy == NULL) {
+                return NULL;
+        }
+
+        proxy->ref_count = 1;
+        proxy->id = ++next_id;
+        proxy->agent = agent_ref(agent);
```

Comment:
This refs the agent, which is a hard ref, but proxy.h comment says this is a weak ref, so either of these are wrong.

I think it needs to be a weak ref actually, because otherwise there is a refcount cycle. I.e. as long as you create any kind of proxy it will keep the agent alive. I.e. you can't unref your initial refcount of the agent and have it destroy any outstanding proxies.

Obviously in the specific case of the agent object this will never be unref:ed in the lifetime of the program, but it is still nice to get these things right. What you want is:

* Main owns the main agent ref.
* Agent owns the proxies via the list (i.e. if you remove from list you unref)
* Proxy has a weak ref (i.e. pointer, but not ref:ed) back to the agent

This means in theory the proxy can outlive its agent, so when removing the proxy from the agent we should probably NULL out ->agent, etc.